scripts/utils/options.ts

```diff
@@ -0,0 +1,245 @@
+): MaybeOptionValues<TOptions> {
+  Object.entries(options)
+    .reduce((acc, [key, option]) => {
+      const flags = optionFlags(key, option);
+
+      if (isBooleanOption(option)) return acc.option(flags, option.description, !!option.inverse);
+
+      const checkStringValue = (raw: string) => {
+        if (!option.values.includes(raw))
+          throw new Error(`Unexpected value '${raw}' for option '${key}'`);
+        return raw;
+      };
+
+      if (isStringOption(option))
+        return acc.option(flags, option.description, (raw) => checkStringValue(raw));
+
+      if (isStringArrayOption(option)) {
+        return acc.option(
+          flags,
+          option.description,
+          (raw, values) => [...values, checkStringValue(raw)],
+          []
+        );
+      }
+
+      throw new Error(`Unexpected option type '${key}'`);
+    }, command)
+    .parse(argv);
+
+  // Note the code above guarantees the types as they come in, so we cast here.
+  // Not sure there is an easier way to do this
+  return command.opts() as MaybeOptionValues<TOptions>;
+}
+
+export function areOptionsSatisfied<TOptions extends OptionSpecifier>(
+  options: TOptions,
+  values: MaybeOptionValues<TOptions>
+) {
+  return !Object.entries(options)
+    .filter(([, option]) => isStringOption(option) && option.required)
+    .find(([key]) => !values[key]);
+}
+
+export async function promptOptions<TOptions extends OptionSpecifier>(
+  options: TOptions,
+  values: MaybeOptionValues<TOptions>
+): Promise<OptionValues<TOptions>> {
+  const questions = Object.entries(options).map(([key, option]): PromptObject => {
+    if (!isBooleanOption(option)) {
+      const currentValue = values[key];
+      return {
+        type: isStringArrayOption(option) ? 'autocompleteMultiselect' : 'select',
+        message: option.description,
+        name: key,
+        // warn: ' ',
+        // pageSize: Object.keys(tasks).length + Object.keys(groups).length,
+        choices: option.values.map((value) => ({
+          title: value,
+          value,
+          selected:
+            currentValue === value ||
+            (Array.isArray(currentValue) && currentValue.includes?.(value)),
+        })),
+      };
+    }
+    return {
+      type: 'toggle',
+      message: option.description,
+      name: key,
+      initial: option.inverse,
+      active: 'yes',
+      inactive: 'no',
+    };
+  });
+
+  const selection = await prompts(questions);
+  // Again the structure of the questions guarantees we get responses of the type we need
+  return selection as OptionValues<TOptions>;
+}
+
+function getFlag<TOption extends Option>(
+  key: OptionId,
+  option: TOption,
+  value?: OptionValue<TOption>
+) {
+  if (isBooleanOption(option)) {
+    const toggled = option.inverse ? !value : value;
+    return toggled ? `--${longFlag(key, option)}` : '';
+  }
+
+  if (isStringArrayOption(option)) {
+    // I'm not sure why TS isn't able to infer that OptionValue<TOption> is a
```

Comment:
I am probably wrong but I think the type inference only goes from the specific type down to its properties, it won't evaluate other types that share the same generic. At the time `value?: OptionValue<TOption>` was declared, `TOption` meant one thing, and after the `if` statement it's another thing.

So, if you used this instead, for instance:
```
(value as OptionValue<typeof option>).map((v) => `--${longFlag(key, option)} ${v}`).join(' ')
```

Then `value` will take the correctly inferred `option` type into consideration, then you would get what you expect

Reply:
I think you are right in the first part, but I also think I tried the second thing and it didn't work either :cry:)

Reply:
Hmm just tired it again and it works. You are right in all things.